### PR TITLE
Update `_unpack_to_numpy` function to convert JAX and PyTorch arrays to NumPy

### LINKED ISSUE #25882: [Bug]: plt.hist takes significantly more time with torch and jax arrays
### Bug summary

Hi,

Time taken to plot `plt.hist` directly on `jax` or `torch` arrays is significantly more than **combined time taken to first convert them to `numpy` and then using `plt.hist`**. Shouldn't `matplotlib` internally convert them to `numpy` arrays before plotting?

To reproduce the bug, directly run the following snippet on Google Colab.

### Code for reproduction

```python
from time import time
import numpy as np

import torch

import jax
import jax.random as jr
import jax.numpy as jnp

import matplotlib.pyplot as plt

jax_array = jr.normal(jr.PRNGKey(0), (1000, 150))
torch_array = torch.randn(1000, 150)

def plot_hist(array):
    init = time()
    plt.figure()
    plt.hist(array)
    print(f"Time to plot: {time() - init:.2f} s")
    plt.show()
    
plot_hist(jax_array.ravel())
plot_hist(torch_array.ravel())
plot_hist(np.array(jax_array.ravel()))
plot_hist(np.array(torch_array.ravel()))
```

### Actual outcome

Time to plot: 4.19 s
image

Time to plot: 2.61 s
image

Time to plot: 0.03 s
image

Time to plot: 0.04 s
image

### Expected outcome

Time to plot: 0.03 s

Time to plot: 0.04 s

Time to plot: 0.03 s

Time to plot: 0.04 s

### Additional information

> What are the conditions under which this bug happens? input parameters, edge cases, etc?

It is happening with all kinds of shapes.

> Has this worked in earlier versions?

Tested with default colab `matplotlib` version 3.7.1 and also with 3.6.3.

> Do you know why this bug is happening?

Not exactly sure.

> Do you maybe even know a fix?

Maybe convert any python object to a `numpy` array before plotting?

### Operating system

Ubuntu 20.04.5 LTS

### Matplotlib Version

3.7.1

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

3.10.11

### Jupyter version

6.4.8

### Installation

None

### PR DESCRIPTION
**Pull request by patel-zeel**
## PR summary [Testing in process]

This PR closes #25882 by modifying the `_unpack_to_numpy` function. The main changes are the following.
* Added an `if` condition to check if an object has `__array__` method and the new object returned by accessing `__array__` method is a NumPy array.
* Added an `if` condition to capture NumPy scalars which were not captured by the `ndarray` check earlier. This was needed because otherwise NumPy scalar objects get infinitely stuck into `__array__` check since they get converted to `ndarray` upon calling `__array__` method on them.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by jklymak** [comment]
Please remove the unrelated changes.

**Comment by oscargus** [comment]
Hmm, it seems like this is probably the clearest indication of the test failures:
```
>       assert ax1.xaxis.get_units() == ax2.xaxis.get_units() == "hours"
E       AssertionError: assert None == 'hours'
```
As the `Quantity` test class has an `__array__` method
https://github.com/matplotlib/matplotlib/blob/10e8bf134c77e73863c65a1f246bb8d9a684fc59/lib/matplotlib/tests/test_units.py#L41-L42
the units will be dropped as the test fail.

I have no idea about the unit support though, so not really clear how to get around it...

**Comment by mwaskom** [comment]
FYI I don't think you're supposed to call [`__array__`](https://pandas.pydata.org/docs/reference/api/pandas.Series.__array__.html
) directly:

> Users should not call this directly. Rather, it is invoked by [numpy.array()](https://numpy.org/doc/stable/reference/generated/numpy.array.html#numpy.array) and [numpy.asarray()](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html#numpy.asarray).

But not sure how much it matters in practice.

**Comment by tacaswell** [comment]
This is hitting the same reason why we do not blindly call `np.asanyarray(...)` on all of our input because that can strip off unit information (which will break some of our users, changing that is off the table for now).

Between this and the note about not calling `__array__` directly probably means this particular solution is not the right path.

**Comment by patel-zeel** [comment]
Thank you for a diverse set of feedback, @oscargus, @mwaskom and @tacaswell. So, what do you think would be a better way to go ahead?

1. Implement `to_numpy()` method in JAX and PyTorch.
2. JAX and PyTorch specific `if` check in `_unpack_to_numpy` function:
```py
if str(type(x)) == "<class 'torch.Tensor'>":
    return x.numpy()
if str(type(x)) == "<class 'jaxlib.xla_extension.ArrayImpl'>":
    return np.asarray(x)
```

**Comment by oscargus** [comment]
Do not trust this fully, but I think that checking if there is a `numpy` method and then calling that (and checking the output) would probably make sense and solve `torch.Tensor`. But I also think that from Matplotlib's perspective (and probably many other libraries), having a well defined interface, like `to_numpy`, on everything that can be converted to NumPy would make most sense.

**Comment by tacaswell** [comment]
I am very :-1: on added string checking of types.

Unfortunately we are in a awkward bind where we very permissive in what we take as input, do not want to depend on any imports, and due to the diversity of input we can not treat them all the same.

**Comment by timhoffm** [comment]
I have not checked, but maybe there is something in the [python array API standard](https://data-apis.org/array-api/2022.12/API_specification/index.html) - At least it would belong there.

**Comment by timhoffm** [comment]
Would [__array_namespace__](https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.array.__array_namespace__.html) a solution for us?

**Comment by jakevdp** [comment]
There's a third option not mentioned here: use `__array__` as the standard convert-to-array function, and give libraries where this is not the apropriate behavior some way to opt-out. Adding a new standard (`to_numpy`) with poorly-defined semantics is not a great way forward in my opinion.

**Comment by mwaskom** [comment]
My sense is that the direction data libraries would like to move is for exchange via the `__array__` method, so despite throwing some cold water on that above, it's probably a better path than trying to get torch to add `to_numpy` in addition to `numpy` and `__array__`.

**Comment by jklymak** [comment]
We discussed the history of this a bit on the dev call today.  I think the below is close to correct, but I could be misunderstanding:

We officially support numpy arrays as inputs to our data plotting functions.

We also officially support mechanisms for objects to get passed that contain "unit" information (eg pint).  Somewhat confusingly, this unit information is sometimes at the container level (eg pint), and sometimes at the element level, or the dtype of the elements (eg nparrays of datetime64, or lists of strings).  

We _unofficially_ support xarray and pandas objects, assume they have no units, by calling their `values` or `to_numpy` methods.  

At the level that `_unpack_to_numpy` is called, we cannot strip units from objects with units, because they have not been checked for yet.  In the case discussed here, it is indeed the unit checking that is slowing things down.  

After we have checked for units, we usually call `np.asarray`.  But we can't call that right away because of our unit support.  

I'm not sure what the path out of the conundrum is - I somewhat feel the unit conversion interface should have been less magical, and more explicit, so users would have to specify a converter on an axis manually, rather than us guessing the converter.

**Comment by rgommers** [comment]
> Would [`__array_namespace__`](https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.array.__array_namespace__.html) be a solution for us?

That isn't quite the right thing; the array API standard is meant to use "native functions", so this method is what you'd use if you want to retrieve the `torch` namespace and use `torch.asarray` & co. Here you specifically want `numpy` arrays instead.

I agree with @jakevdp and @mwaskom that use of `__array__` is more idiomatic. The most standard thing is `np.asarray` (which relies on `__array__`, or the Python buffer protocol, or DLPack), but if that's too permissive than using `__array__` directly is fine.



> After we have checked for units, we usually call `np.asarray`. But we can't call that right away because of our unit support.

If units libraries silently lose data when `np.asarray` is called on their container objects, they really should implement `__array__` and make it raise an exception. This is also what, for example, sparse arrays do.

**Comment by timhoffm** [comment]
From an interface perspective, it's reasonable to rely on `__array__`. I think we should investigate how we can make this work internally.

**Comment by patel-zeel** [comment]
@timhoffm It has been a while since I worked on this PR. Can you please suggest if your latest suggestion in #25882 will resolve the issue we are discussing in this PR? If not, could you please suggest potential workarounds?

**Comment by timhoffm** [comment]
@patel-zeel in my comment https://github.com/matplotlib/matplotlib/issues/25882#issuecomment-1872440671, I hadn't considered the unit problem. That indeed makes the problem much more complicated.

To all: To summarize and comment on the above proposed solutions:

1. **Implement `to_numpy()` method in JAX and PyTorch**
   IMO this won't happen. With some right, they say that `__array__` is nowadays their ideomatic hook to turn them into numpy arrays.
2. **Import and type-check by type**
   We don't want to try and import complex libraries just because someone might have passed an element of that type. This would be a performance hit for all users that happen to have that library installed, but don't use it.
3. **type-check by string**
   This is inelegant and brittle.
4. **Rework our unit handling system**
   Any changes to unit handling, that could help here, would definitively be a major project in matplotlib, and would likely also require changes for some downstream users.

There is no easy solution here. Special situations sometimes require special measures:

Given all the boundary conditions, __I'd be +0.5 on *type-checking by string*__, despite @tacaswell being strongly 👎 on this. Usually, I'd agree, but that's the only realistic way forward. 1. won't happen; 2. is introducing strong coupling, which IMHO is worse; 4. won't realistcally happen, because we don't have the capacity for it.

So what would we buy into with *type-checking by string*. Drawbacks are (1) the str comparison is slower than a type check - but that should be negligible; and (2) It's brittle because the str representation could change without us noticing and then the functionality would be broken. To alleviate (2), we could use `f"{type(x).__module__}.{type(x).__qualname__}"` That leaves out unnecessary fluff and would only change when the libraries reorganize. Additionally, in the worst case scenario that the string changes, we would fall back to the current solution.
In short: we can easily make using JAX/Torch arrays faster with the string-type check; With a not too high likelihood, that can break in the future, which would bring us back to where we are now. - Sounds like a reasonable deal to me.

The only other alternative would be to tell users to convert their JAX/Torch arrays explicitly (or live with the performance impact). But that'd be not user friendly.

**Comment by tacaswell** [comment]
>  4. won't realistcally happen, because we don't have the capacity for it.

This is what Kyle is working on, but is 1-2 years off, but I don't think we should wait for it.

----

I am convinced by @timhoffm 's analysis and am also +0.5 on string typing now.

**Comment by jakevdp** [comment]
Couldn’t you accomplish option 2 without the performance impact by looking to see if certain modules are already in `sys.modules`?

**Comment by timhoffm** [comment]
That may indeed work. e.g. `“torch in sys.modules’ is very likely to be true if a user can create and pass to us a torch tensor.  It‘s not a strict prerequisite, but the the edge cases would require non-canonical imports (e.g. `import torch as _torch`) or messing with the import system. Such cases would not be detected and fall back to the current implementation.

To be clear: This is viable here, because it‘s works under normal circumstances (98%), and the missing cases are covered by the current fallback.

**Comment by patel-zeel** [comment]
@jakevdp @timhoffm Are you suggesting something like this?

```py
if "torch" in sys.modules:
    from torch import Tensor as TorchTensor
    if isinstance(x, TorchTensor):
        return x.numpy()
if "jax" in sys.modules:
    from jax import Array as JAXArray
    if isinstance(x, JAXArray):
        return x.__array__()    
```

**Comment by jakevdp** [comment]
Yeah, probably something like that. I might even avoid the imports, and wrap with a `try`/`except` just in case there's something strange (like `torch.py` in the namespace, which either doesn't define `Tensor` or defines `Tensor` as a non-type):
```python
import sys

def is_torch_array(x) -> bool:
  try:
    return isinstance(x, sys.modules['torch'].Tensor)
  except:
    return False

def is_jax_array(x) -> bool:
  try:
    return isinstance(x, sys.modules['jax'].Array)
  except:
    return False
```
I think the only way to confound this would be if the user manually deleted entries from `sys.modules`, in which case the worst thing that would happen is a false-negative output.

File: lib/matplotlib/cbook.py [hunk 1]
```diff
@@ -2234,6 +2235,28 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
     return cls.__new__(cls)
 
 
+def _is_torch_array(x):
+    try:
+        return isinstance(x, sys.modules['torch'].Tensor)
+    except (KeyError, AttributeError, TypeError):
```
Comment (jakevdp) [inline lib/matplotlib/cbook.py:2241]: I think a raw except would be safer here: for example, if someone defines a file named `torch.py` or `jax.py` that has a non-trivial module-level `__getattr__`, this could raise a `ValueError` or something else, and it would make `matplotlib` unusable with no real recourse. I think a raw`except` is preferable to that.

Comment (jakevdp) [reply lib/matplotlib/cbook.py:2241]: Oh, just saw the warning below. I'm -1 on handing raw exceptions with a warning. We should just silently return `False` for all exceptions.

**Comment by patel-zeel** [comment]
@jakevdp Thanks for the improved solution! I'd have never though about `torch.py` in namespace and also various exceptions in different scenarios.

@timhoffm I have applied @jakevdp's solution. I am now wondering how to test this feature or should we even test it or not.

File: lib/matplotlib/cbook.py (hunk 1 above)
Comment (patel-zeel) [reply lib/matplotlib/cbook.py:2241]: @jakevdp I totally agree! For this case, I'd detach myself from "Errors should never pass silently." from "The Zen of Python, by Tim Peters" :)

File: lib/matplotlib/cbook.py [hunk 2]
```diff
@@ -2234,6 +2234,22 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
     return cls.__new__(cls)
 
 
+def _is_torch_array(x):
+    """Check if 'x' is a PyTorch Tensor."""
+    try:
```
Comment (timhoffm) [inline lib/matplotlib/cbook.py:2239]:
```suggestion
    # we're intentionally not attempting to import torch. If somebody
    # has created a torch array, torch should already be in sys.modules
    try:
```

let's document this for our future selves. Please also add a similar one to _ is_ jax_array.

File: lib/matplotlib/cbook.py [hunk 3]
```diff
@@ -2234,6 +2234,22 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
...
+        return isinstance(x, sys.modules['torch'].Tensor)
+    except Exception as e:  # not using bare `except` to bypass flake8
```
Comment (timhoffm) [inline lib/matplotlib/cbook.py:2241]:
```suggestion
    except Exception as e:
```

This is not to placate flake8 but *just* the correct way. If you want to catch "all errors", `except Exception` is almost always the way to go. From https://docs.python.org/3/library/exceptions.html#Exception

> All built-in, non-system-exiting exceptions are derived from this class.

We don't want to catch `KeyboardInterrupt` or `SystemExit`, which would be caught by a bare except.

File: lib/matplotlib/cbook.py [hunk 4]
```diff
@@ -2234,6 +2234,22 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
...
+        return False
+
+
+def _is_jax_array(x):
+    """Check if 'x' is a JAX Array."""
+    try:
+        return isinstance(x, sys.modules['jax'].Array)
+    except Exception as e:  # not using bare `except` to bypass flake8
```
Comment (timhoffm) [inline lib/matplotlib/cbook.py:2249]:
```suggestion
    except Exception as e:
```

as above

**Comment by timhoffm** [comment]
> I am now wondering how to test this feature or should we even consider testing it.

Testing this is quite hard because we don't have PyTorch or JAX as test dependencies. The only possibility I see is mocking. The mock would be quite implementation specific (make a namespace with a Tensor class with an `__array__` attribute). So we cannot really test whether _is_torch_array actually detects torch arrays. But what is still valuable is passing such a mock to `_unpack_to_numpy` and testing that the numpy array from `__array__` is retuned.

**Comment by patel-zeel** [comment]
@timhoffm Thanks for the review and suggestion for testing. I have applied the suggested changes and implemented the first version of testing this feature.

File: lib/matplotlib/tests/test_cbook.py [hunk 5]
```diff
@@ -938,3 +940,45 @@ def test_auto_format_str(fmt, value, result):
...
+            return self.data
+    torch = ModuleType('torch')
+    torch.Tensor = Tensor
+    sys.modules['torch'] = torch
+
+    data = np.arange(10)
+    torch_tensor = torch.Tensor(data)
+
+    result = cbook._unpack_to_numpy(torch_tensor)
+    assert isinstance(result, np.ndarray)
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_cbook.py:962]: I would be more specific: `_unpack_to_numpy` should not just return any array, but exercise the `__array__` interface
```suggestion
    assert result is torch_tensor.__array__()
```

Note that formally, we would want
```
    nptest.assert_array_equal(_unpack_to_numpy(torch.Tensor(data)), data)
```
which should work for any array wrapper, but since we're mocking anyway I think it's justifiable to test the mocked implementation detail.

File: lib/matplotlib/tests/test_cbook.py [hunk 6]
```diff
@@ -938,3 +940,45 @@ def test_auto_format_str(fmt, value, result):
...
+    jax = ModuleType('jax')
+    jax.Array = Array
+
+    sys.modules['jax'] = jax
+
+    data = np.arange(10)
+    jax_array = jax.Array(data)
+
+    result = cbook._unpack_to_numpy(jax_array)
+    assert isinstance(result, np.ndarray)
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_cbook.py:984]: As above:
```suggestion
    assert result is jax_array.__array__()
```

File: lib/matplotlib/tests/test_cbook.py (hunk 5 above)
Comment (patel-zeel) [reply lib/matplotlib/tests/test_cbook.py]: I agree!

File: lib/matplotlib/cbook.py [hunk 7]
```diff
@@ -2344,6 +2344,26 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
     return cls.__new__(cls)
 
 
+def _is_torch_array(x):
+    """Check if 'x' is a PyTorch Tensor."""
+    try:
+        # we're intentionally not attempting to import torch. If somebody
+        # has created a torch array, torch should already be in sys.modules
+        return isinstance(x, sys.modules['torch'].Tensor)
+    except Exception:
```
Comment (jakevdp) [inline lib/matplotlib/cbook.py:2353]: I would add a note here about why we're using raw `Exception`, because someone will question this in the future!

**Review by jakevdp — COMMENTED** [review]
Looks good – one comment is that perhaps we should abstract this a bit: maybe have a configurable list of external objects to look for (e.g. `external_objects = ['torch.Tensor', 'jax.Array']`) and write just a single function that loops through and checks these.

Then in the future, if someone wanted to add `cupy.ndarray` or something it would be just a tiny change, and it could even be done at runtime if we wanted to provide that API.

File: lib/matplotlib/cbook.py (hunk 7 above)
Comment (patel-zeel) [reply lib/matplotlib/cbook.py:2353]: @jakevdp That's true. How about the following note?

```py
except Exception:
    # we're catching all exceptions intentionally during isinstance check to 
    # prevent potential unknown exceptions and avoid `matplotlib` crashes.
    return False
```

Comment (jakevdp) [reply lib/matplotlib/cbook.py:2353]: I'd say something more specific, like `we're attempting to access attributes on imported modules which may have arbitrary user code, so we deliberately catch all exceptions`.

**Comment by patel-zeel** [comment]
@jakevdp How'd you suggest abstracting this? Some relevant points:
* I tried creating a `cupy` array and getting the NumPy array with `__array__()` method but it breaks and suggests to use `.get()` method.
```py
import cupy
array = cupy.array([1, 2, 3.0])
np_array = array.__array__()  # fails
# np_array = array.get()  # this works
```
Output:
```py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], [line 1](vscode-notebook-cell:?execution_count=3&line=1)
----> [1](vscode-notebook-cell:?execution_count=3&line=1) array.__array__()

File cupy/_core/core.pyx:1475, in cupy._core.core._ndarray_base.__array__()

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```
* I came across [this issue](https://github.com/pytorch/pytorch/issues/20778) on PyTorch and realized that `.numpy(force=True)` method can help in cases where arrays need `.detach()` and/or `.cpu()` before `.__array__()` can successfully get the underlying NumPy array. As far as `matplotlib` is concerned, I guess `.numpy(force=True)` can be a better alternative compared to `.__array__()` for PyTorch (or even `.detach().cpu().numpy()` can work to support the older versions).


Considering both of the above cases, discussion in #25882, and discussion in this PR, would it be better to provide two methods, `is_type` and `to_numpy`, for each object like the following?

```py
import sys
import numpy as np
from abc import ABC, abstractmethod

class TypeArray(ABC):
    @abstractmethod
    def is_type(x):
        pass
    
    @abstractmethod
    def to_numpy(x):
        pass

class TorchArray(TypeArray):
    def is_type(x):
        """Check if 'x' is a PyTorch Tensor."""
        try:
            # we're intentionally not attempting to import torch. If somebody
            # has created a torch array, torch should already be in sys.modules
            return isinstance(x, sys.modules['torch'].Tensor)
        except Exception:  # TypeError, KeyError, AttributeError, maybe others?
            # we're attempting to access attributes on imported modules which
            # may have arbitrary user code, so we deliberately catch all exceptions
            return False
    def to_numpy(x):
        """Convert to NumPy array"""
        # preferred over `.numpy(force=True)` to support older PyTorch versions.
        return x.detach().cpu().numpy()  
    
class JaxArray(TypeArray):
    def is_type(x):
        """Check if 'x' is a JAX array."""
        try:
            # we're intentionally not attempting to import jax. If somebody
            # has created a jax array, jax should already be in sys.modules
            return isinstance(x, sys.modules['jax'].Array)
        except Exception:  # TypeError, KeyError, AttributeError, maybe others?
            # we're attempting to access attributes on imported modules which
            # may have arbitrary user code, so we deliberately catch all exceptions
            return False
    def to_numpy(x):
        """Convert to NumPy array"""
        return x.__array__()  # works even if `x` is on GPU
    
class CupyArray(TypeArray):
    def is_type(x):
        """Check if 'x' is a CuPy array."""
        try:
            # we're intentionally not attempting to import cupy. If somebody
            # has created a cupy array, cupy should already be in sys.modules
            return isinstance(x, sys.modules['cupy'].ndarray)
        except Exception:  # TypeError, KeyError, AttributeError, maybe others?
            # we're attempting to access attributes on imported modules which
            # may have arbitrary user code, so we deliberately catch all exceptions
            return False
    def to_numpy(x):
        """Convert to NumPy array"""
        return x.get()

external_objects = [TorchArray, JaxArray, CupyArray]

def _unpack_to_numpy(x):
    """Internal helper to extract data from e.g. pandas and xarray objects."""
    if isinstance(x, np.ndarray):
        # If numpy, return directly
        return x
    if hasattr(x, 'to_numpy'):
        # Assume that any to_numpy() method actually returns a numpy array
        return x.to_numpy()
    if hasattr(x, 'values'):
        xtmp = x.values
        # For example a dict has a 'values' attribute, but it is not a property
        # so in this case we do not want to return a function
        if isinstance(xtmp, np.ndarray):
            return xtmp
    
    for obj in external_objects:
        if obj.is_type(x):
            xtmp = obj.to_numpy(x)
            
            # In case to_numpy() doesn't return a numpy array in future
            if isinstance(xtmp, np.ndarray):
                return xtmp
    return x
```

**Comment by timhoffm** [comment]
IMHO further abstraction would be premature. The current implementation is simple and good enough. Paraphrased from https://youtu.be/UANN2Eu6ZnM?feature=shared

> If something happens for the first time, do a concrete implementation. If it happens for the second time, copy andcadaptvrhe code. If it happens for the third time, factor out commonalities.

This has two major advantages: 1. You don't create abstractions that you don't use. 2. When you build the abstraction, you have three concrete use cases, so it's more likely the abstraction is suitable.

**Review by timhoffm — APPROVED** [review]

**Comment by jakevdp** [comment]
I didn't mean to suggest any complicated abstraction; I was thinking something simple like this:
```python
ARRAYLIKE_OBJECTS = [('jax', 'Array'), ('torch', 'Tensor')]

def maybe_convert_to_array(x):
  for mod, name in ARRAYLIKE_OBJECTS:
    try:
      if isinstance(x, getattr(sys.modules[mod], name)):
        return x.__array__()
    except Exception:
      pass
  return x
```

File: lib/matplotlib/cbook.py [hunk 8]
```diff
@@ -2358,6 +2382,12 @@ def _unpack_to_numpy(x):
         # so in this case we do not want to return a function
         if isinstance(xtmp, np.ndarray):
             return xtmp
+    if _is_torch_array(x) or _is_jax_array(x):
+        xtmp = x.__array__()
```
Comment (jakevdp) [inline lib/matplotlib/cbook.py:2386]: Is there any reason not to do `xtmp = np.asarray(x)` here?

**Comment by timhoffm** [comment]
> I didn't mean to suggest any complicated abstraction; I was thinking something simple like this:
> 
> [...]
> 
> It reduces duplication of logic and makes it easier to add additional types if/when needed. If you wanted to add cupy support, it would just require doing `ARRAYLIKE_OBJECTS.append(('cupy', 'ndarray'))`

Yes, that would be marginally better, and can optionally be done. In the interest of not endlessly bikeshedding the PR, I have accepted the current version. After all, this is all internal and can be refactored any time.

File: lib/matplotlib/cbook.py (hunk 8 above)
Comment (timhoffm) [reply lib/matplotlib/cbook.py:2386]: I think these are equivalent for the handled cases. While [pandas claims](https://pandas.pydata.org/docs/reference/api/pandas.Series.__array__.html#pandas.Series.__array__) you shouldn't call `__array__` directly, I haven't found any official recommendation for it in numpy (which defines the `__array__` API).

For me, either works. `__array__` is more explicit, which can be a good thing, but might be too low level. OTOH I don't think `np.asarray()` will change the implementation so that the `asarray` abstraction would be safer.

Comment (patel-zeel) [reply lib/matplotlib/cbook.py:2386]: @jakevdp @timhoffm What's your opinion about using `x.numpy(force=True)` or `x.detach().cpu().numpy()` for PyTorch? As a user, I'd find this change useful for day-to-day coding since it saves me from manually writing it for every array I want to plot with matplotlib.

**Comment by jklymak** [comment]
OK, apologies for not paying attention to this properly, but hard-coding certain libraries to have a cut around seems incorrect and brittle to me.    What cutoff will we have if we get requests to support other libraries?  

I think the fundamental problem is with where `cbook._reshape_2D` gets called.  This method is only used for `hist` and a couple of the helpers for `violin_stats` and `boxplot_stats`.  

I think it would be a mistake to change `_unpack_to_numpy`, which is used for all unit conversion, and hence almost every other plot method, to work around this problem.  If it were me, I'd split the preprocessing in `hist` and friends to keep units where they are required (eg using `Quantity`), and strip them when we are ready to do so.  

The methods where this gets used are all binning methods.  The `bins` need to keep units (so they can be added to the axes properly).  However, the data needs to be turned into an array to pass to `histogram` (or other stats functions).  I think the proper solution here is to properly differentiate these roles in `hist` (and friends).

**Comment by timhoffm** [comment]
> hard-coding certain libraries to have a cut around seems incorrect and brittle to me.

This is indeed a workaround. The proper way would be for `_unpack_to_numpy()` to use the `__array__` interface for all types if available (maybe through `np.asarray()`). However, in the current internal usages this may lead to loss of units. If you have an alternate proposal how to make JAX and Pytorch arrays work, I'm more than happy to take that. - t.b.h. I don't fully oversee the unit handling and its implications.

Otherwise, I think this PR is good enough to be included in 3.9. It achieves the desired speedup and otherwise is completely internal, so we can still change the implementation whenever we like.

> What criteria will we have if we get requests to support other libraries?

Case-by-case. Support them if it's easily possible, don't if it's not. There's little maintanance burden and no API liability. Also, I don't expect that there would be more than a hand full of such libraries.

File: lib/matplotlib/cbook.py (hunk 8 above)
Comment (timhoffm) [reply lib/matplotlib/cbook.py:2386]: Seems reasonable to use the `x.numpy` method provided by torch. Though I have to say, I don't know whether `__array__` would do something different.

Comment (patel-zeel) [reply lib/matplotlib/cbook.py:2386]: @timhoffm To add more context, in JAX, `.__array__()` method converts a JAX array to a NumPy array irrespective of whether the JAX array is on CPU, GPU, TPU and probably other hardware accelerators. OTOH in PyTorch, they don't do it by default due to uncertainty about performance impact (full discussion is in [this issue](https://github.com/pytorch/pytorch/issues/20778)). So, when we do `x.numpy(force=True)`, it forcefully converts it to NumPy irrespective of the device of the array (handles few others cases as well, e.g. if array has a computation graph for backpropagation then `x.detach()` needs to be called first). I am not sure if this should be handled in a separate issue or we can use `x.numpy(force=True)` in this PR itself but I'm sure that PyTorch users would love this change to avoid writing `x.detach().cpu().numpy()` every time they plot a PyTorch array.

cc @jakevdp.

Comment (jklymak) [reply lib/matplotlib/cbook.py:2386]: My hesitation about this is that memory overflows will get shunted to Matpolotlib, and then we will get the bug reports, whereas if people did `x.*.numpy()` in their own code, they would see what the problem is.  Jax arrays can be far larger than memory allows, and Matplotlib blindly unpacking them for naive users seems like a bad idea.

Comment (rgommers) [reply lib/matplotlib/cbook.py:2386]: > Though I have to say, I don't know whether `__array__` would do something different.

`__array__ is the last thing tried by NumPy when a user would call `np.asarray`, after the buffer protocol and `__array_interface__` (as documented for example [here](https://numpy.org/devdocs/user/basics.interoperability.html#using-arbitrary-objects-in-numpy)). So they're not equivalent, and calling `np.asarray` is idiomatic.

**Comment by patel-zeel** [comment]
@timhoffm @jakevdp Getting back to this after a while. To summarize the pending changes:

* I guess `.numpy(force=True)` for PyTorch can go into silent memory overflows, so we may avoid doing that and let users do it manually if needed.
* Simple abstraction suggested by @jakevdp seems great but `np.asarray(x)` doesn't work for `cupy` so I think that'd be hard to generalize just yet.

I think the accepted changes are optimal based on the current circumstances.

File: lib/matplotlib/cbook.py [hunk 9]
```diff
@@ -2344,6 +2344,30 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
     return cls.__new__(cls)
 
 
+def _is_torch_array(x):
+    """Check if 'x' is a PyTorch Tensor."""
+    try:
+        # we're intentionally not attempting to import torch. If somebody
+        # has created a torch array, torch should already be in sys.modules
+        return isinstance(x, sys.modules['torch'].Tensor)
+    except Exception:  # TypeError, KeyError, AttributeError, maybe others?
```
Comment (tacaswell) [inline lib/matplotlib/cbook.py:2353]: I think this is too permissive, can we do this as 

```python
if sys.modules.get('torch', None) is not None:
     return isinstance(x, sys.modules['torch'].Tensor
```

and similar for jax?

Comment (tacaswell) [reply lib/matplotlib/cbook.py:2353]: or

```python
if (mod := sys.modules.get('torch', None)) is not None:
    return isinstance(x, mod.Tensor)
```

Comment (jakevdp) [reply lib/matplotlib/cbook.py:2353]: The failure mode there is that if someone names a file `jax.py` or `torch.py`, then matplotlib can no longer be used.

Comment (tacaswell) [reply lib/matplotlib/cbook.py:2353]: fair point.  I retract my suggestion.

**Comment by tacaswell** [comment]
I'm going to merge this to move forward.  On one hand, I think it is reasonable to expect users to get their data back to the cpu and in numpy before we plot it, but we have gotten enough bugs and this is a light enough touch. 

If using `__array__()` directly is something we really should not do, we can do a follow on tweaking the exact behavior for jax / torch.

This also sets a reasonble pattern for how we would add support for cupy / the next big library.

**Comment by tacaswell** [comment]
Thank you for your work on this @patel-zeel and congratulations on your first merged Matplotlib PR 🎉 

I hope we hear from you again.

**Comment by patel-zeel** [comment]
Thank you, @tacaswell, and all the contributors to this PR. This PR has taught me a lot of dev tricks. This wouldn't be possible without everybody's diverse viewpoints, @jakevdp's robust ideas, and @timhoffm's pivotal role.
